src/Highlight.ts

```diff
@@ -91,17 +100,48 @@ export default class Highlight {
   }
 
   public serialize(referenceElement?: HTMLElement): SerializedHighlight {
-    referenceElement = referenceElement || dom(this.range.commonAncestorContainer).closest('[id]');
+    const validReferenceElement = this.getValidReferenceElement(
+      this.options.skipIDsBy, referenceElement
+    );
+
+    if (!validReferenceElement) {
+      throw new Error('reference element not found');
+    }
 
     return new SerializedHighlight({
       ...this.data,
-      ...SerializedHighlight.defaultSerializer(this.range, referenceElement),
+      ...SerializedHighlight.defaultSerializer(this.range, validReferenceElement),
     });
   }
   private loadStyle() {
-    const {style} = this.data;
+    const { style } = this.data;
     if (style) {
       this.elements.forEach((element) => element.classList.add(style));
     }
   }
+
+  private checkReferenceElement(validatingRegexps: RegExp[], referenceElement?: HTMLElement): boolean {
```

Comment:
```suggestion
  private checkReferenceElement(validatingRegexps: RegExp[], referenceElement?: HTMLElement): referenceElement is HTMLElement {
```

will turn this into a guard so that when it passes ts will know its an html element and not undefined, and you can get rid of your `as HTMLElement` on line 139

Reply:
But there is a chance that the `referenceElement` is a `HTMLElement` but it's not a valid one in which case i would need to access its parent. Am I missing something?

Reply:
my suggested change doesn't affect the conditions that you wrote for when the function returns true or false, it just says that when it _does_ return true, typescript will know that it is not undefined